### PR TITLE
Dev sentinel securityinsights microsoft.security insights 2024 04 01 preview

### PR DESCRIPTION
**Pull request by xuhumsft**
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

> [!NOTE]
> As of January 2024 there is no PR assignee. This is expected. See https://aka.ms/azsdk/pr-arm-review.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

diagram

<details>
<summary>Click here to see the details of Step 1, Breaking Changes review</summary>

If you are in purview of Step 1 of the diagram, follow the [Breaking Changes review process].  
**IMPORTANT!** This applies even if you believe your PR was mislabeled, for any reason, including tool failure.
</details>

<details> 
<summary>Click here to see the details of Step 2, ARM review</summary>

See https://aka.ms/azsdk/pr-arm-review.
</details>

<details>
<summary> Click here to see the diagram footnotes</summary>

### Diagram footnotes

[1] See [ARM review queue] (for **PR merge** queues, see [2]).  
[2] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])  
The ARM reviewer on-call engineer visits the merge queue twice a day, so the approximate ETA for merges is 12 - 24 hours.

</details>

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been generated using [OpenAPI Hub](https://aka.ms/openapihub)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [ ] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [ ] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with PR workflow diagram Step 2 (ARM review), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://aka.ms/azsdk/pr-arm-review#about-the-arm-review-queue
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[Breaking Changes review process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

### CODE REVIEW

**Comment by xuhumsft** [comment]
The only newly introduced is business applications agents from https://github.com/Azure/azure-rest-api-specs/pull/28504. The Breaking change and LintDiff is generated from openAPIHub automatically. Please advise what is the next step

**Comment by xuhumsft** [comment]
https://github.com/Azure/azure-rest-api-specs/pull/28504 is already granted with ARMSignedOff label

**Comment by xuhumsft** [comment]
> The only newly introduced is business applications agents from #28504. The Breaking change, LintDiff etc. is generated from openAPIHub automatically. Please advise what is the next step

https://github.com/Azure/azure-rest-api-specs/pull/28504 is already granted with ARMSignedOff label

**Comment by gary-x-li** [comment]
Please get a breaking change review before moving to ARM review (step 2).

Please fill out the **Due diligence checklist** section and the [Avocado Errors](https://github.com/Azure/azure-rest-api-specs/pull/28437/checks?check_run_id=23358212701) after you get the breaking change approval.

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2024-04-01-preview/BusinessApplicationAgents.json
Comment (raosuhas) [inline specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2024-04-01-preview/BusinessApplicationAgents.json:513]: >listActions [](http://example.com/codeflow?start=251&length=11)

Looks like you have modelled these as a bunch of POST actions : listactions, undoactions and reportactionstatus

Any reason why you have not modelled these with PUT\GET\DELETE semantics ?

Also how are these actions trigerred ? is that not a client operation ?

**Comment by raosuhas** [comment]
>           "azureResourceId": "/subscriptions/d0cfe6b2-9ac0-4464-9919-dccaee2e48c0/resourceGroups/myRg/providers/Microsoft.OperationalInsights/workspaces/myWorkspace",

Is this resource id differentfrom the workspace for which you are creating the "systems " resource ? Why do you need this duplication ?

---
Refers to: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2024-04-01-preview/examples/businessApplicationAgents/systems/Systems_CreateOrUpdate.json:14 in b1c80aa. [](commit_id = b1c80aa7a7b8762f97f0163b3a79645e8d90ab37, deletion_comment = False)

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2024-04-01-preview/BusinessApplicationAgents.json
Comment (dvir-ms) [reply specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2024-04-01-preview/BusinessApplicationAgents.json:513]: hi @raosuhas 
thanks for the comment 

indeed - the creation of the action happens internally (as part of automatic mechanism that we have) 
the user is not creating it. 
the user can only list the actions, update their status, and undo (which is similar to update but for another field)

the updates are only support updating specific fields and not the whole instance.

**Comment by dvir-ms** [comment]
> > ```
> >       "azureResourceId": "/subscriptions/d0cfe6b2-9ac0-4464-9919-dccaee2e48c0/resourceGroups/myRg/providers/Microsoft.OperationalInsights/workspaces/myWorkspace",
> > ```
> 
> Is this resource id differentfrom the workspace for which you are creating the "systems " resource ? Why do you need this duplication ?
> 
> Refers to: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2024-04-01-preview/examples/businessApplicationAgents/systems/Systems_CreateOrUpdate.json:14 in [b1c80aa](https://github.com/Azure/azure-rest-api-specs/commit/b1c80aa7a7b8762f97f0163b3a79645e8d90ab37). [](commit_id = [b1c80aa](https://github.com/Azure/azure-rest-api-specs/commit/b1c80aa7a7b8762f97f0163b3a79645e8d90ab37), deletion_comment = False)

**Comment by dvir-ms** [comment]
> Is this resource id differentfrom the workspace for which you are creating the "systems " resource ? Why do you need this duplication ?

hi 

I accidently closed the PR so sorry for that 
regarding your comment - 
in this example indeed it is the same but it can be anything and it's not limited to its workspace or generally to a workspace.

**Comment by raosuhas** [comment]
> ## Next Steps to Merge
> Next steps that must be taken to merge this PR:
> * ❌ Your PR has at least one breaking change (label: `BreakingChangeReviewRequired`). See the PR description for help.
> * ❌ The required check named `Breaking Change(Cross-Version)` has failed. Refer to the check in the PR's 'Checks' tab for details on how to fix it. In addition, refer to step 1 in the [PR workflow diagram](https://github.com/Azure/azure-rest-api-specs/blob/main/.github/PULL_REQUEST_TEMPLATE/control_plane_template.md#pr-review-workflow-diagram) (even if your PR is for data plane, not ARM).

The ARM changes are fine, please follow the steps above to get breaking change approval, before restarting ARM review

**Comment by xuhumsft** [comment]
> > ## Next Steps to Merge
> > Next steps that must be taken to merge this PR:
> > 
> > * ❌ Your PR has at least one breaking change (label: `BreakingChangeReviewRequired`). See the PR description for help.
> > * ❌ The required check named `Breaking Change(Cross-Version)` has failed. Refer to the check in the PR's 'Checks' tab for details on how to fix it. In addition, refer to step 1 in the [PR workflow diagram](https://github.com/Azure/azure-rest-api-specs/blob/main/.github/PULL_REQUEST_TEMPLATE/control_plane_template.md#pr-review-workflow-diagram) (even if your PR is for data plane, not ARM).
> 
> The ARM changes are fine, please follow the steps above to get breaking change approval, before restarting ARM review

Hi @raosuhas. 
We talked in the office hour on Monday. The breaking changes and lint diff is generated by openAPI hub. The only change in this release is the businessApplicationAgents. And it is new resource type. We do modify any existing resource types. Thank you.

**Comment by raosuhas** [comment]
Discussed this again with the team. The breaking changes are against the latest released stable version. Please work with the breaking changes team and get it approved or fix the changes. You can remove "ARMChangesRequested" once the breaking change label has been taken care of.

**Comment by xuhumsft** [comment]
The breaking change are resolved. The preview version is consistent with the stable version.
For the LintDiff, it seems also caused by properties in preview only. This is not changed from previous preview as we can see in the log.

**Comment by TimLovellSmith** [comment]
Please compute 'due diligence checklist'

**Comment by TimLovellSmith** [comment]
Adding suppressions for e.g. 'customDetails' will enable us to 'record' the knowledge that these are not real linter errors for this API, simplifying this and future reviews (once suppression review is done).

**Comment by TimLovellSmith** [comment]
This part of the API looks probably incorrectly defined to me: it looks like its not actually just a 'get', its a 'list' API, according to ARM conventions, right.

"list all the geodatas of the IP enrichment". (shouldn't enrichments be plural? I suppose geodata is arguably already a plural; 'data' vs 'datum')

But the description says it gets, not lists... 

>   "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.SecurityInsights/enrichment/ip/geodata/": {
      "get": {
        "x-ms-examples": {
          "Get geodata for a single IP address": {
            "$ref": "./examples/enrichment/GetGeodataByIp.json"
          }
        },
        "tags": [
          "Enrichment"
        ],
        "description": "Get geodata for a single IP address",
        "operationId": "IPGeodata_Get",

**Comment by TimLovellSmith** [comment]
Aside from what I've mentioned so far there seem to be many lintDiff errors for this new API version, and I don't see them all as justified. Please go through them again.

**Comment by NoaGoren-Microsoft** [comment]
**Entities API, EntityQueries API and EntityQueryTemplates API:**
These APIs rely on Graph entities, which frequently undergo changes in their properties. To accommodate these dynamic changes, additional properties are essential.

**Entities API:**
This API uses additional properties for Entity Edges, which describe relationships between entities. These relationships change frequently and are not defined by a fixed model, necessitating flexible and adaptive properties.

**EntityQueries API:**
This API allows for customer-customized activities. It uses additional properties for the entity filter property, which relies on fields from related entities. Since these fields are undefined in advance and subject to frequent changes, having flexible properties is crucial.

**EntityQueryTemplates API:**
This API lists out-of-the-box activities. Similar to the EntityQueries API, it uses additional properties for entity filters. These filters depend on the dynamic and frequently changing fields of related entities, making flexible properties necessary.

**Comment by sagamzu** [comment]
**DataConnectors API**

The DataConnectors API is a service designed to aggregate data from various APIs into the customer's workspace.
The customer defines the sources and methods for data collection through the DataConnectors API. 
Before saving the resource with the provided instructions, we make a request to the given API (using the Scuba service) to validate the instructions.

The API uses additional properties for different configurations that the customer must define to send a request to the desired endpoint  ("apiEndpoint" property) and retrieve data. 
These properties can be different for each API, and we don't know them in advanced. 

For example:
- authorizationEndpointQueryParameters: The customer provides an authorization endpoint (that require in order to get data from the "apiEndpoint") and any necessary query parameters to access it. Each API may require different query parameters.
- Headers: To access the endpoint, the customer must provide headers. The required headers vary between endpoints and consist of different sets of key-value pairs, depending on the data source.

**Comment by NoaGoren-Microsoft** [comment]
Entities API Lint suppressions:
We got several Lint errors on this API:
1. Property name should be camel case.
2. Get endpoints for collections of resources must only have the value and nextLink properties in their model.
3. A PUT operation request body schema should be the same as its 200 response schema, to allow reusing the same entity between GET and PUT. If the schema of the PUT request body is a superset of the GET response body, make sure you have a PATCH operation to make the resource updatable. Operation: 'EntityQueries_CreateOrUpdate' Request Model: 'parameters[5].schema' Response Model: 'responses[200].schema'

This API has not been changed in the past year. This is an API used by customers and we haven't recieved such errros beforehand, and we cannot change it much.

**Comment by xuhumsft** [comment]
Here is the comments for AvoidAdditionalProperties in ThreatIntelligenceQuery.json from @moranraz

We are planning to add new APIs soon that will specify the exact object schemas, and then we will replace the additional properties with those definitions. For now, we got an exemption from the ARM reviewer to use additionalProperties

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2024-04-01-preview/BusinessApplicationAgents.json
Comment (mentat9) [inline specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2024-04-01-preview/BusinessApplicationAgents.json:882]: >"The secret source of the agent.\r\nAzureKeyVault are the options" [](http://example.com/codeflow?start=25&length=66)

Fix grammar.

**Comment by mentat9** [comment]
> Entities API Lint suppressions: We got several Lint errors on this API:
> 
> 1. Property name should be camel case.
> 2. Get endpoints for collections of resources must only have the value and nextLink properties in their model.
> 3. A PUT operation request body schema should be the same as its 200 response schema, to allow reusing the same entity between GET and PUT. If the schema of the PUT request body is a superset of the GET response body, make sure you have a PATCH operation to make the resource updatable. Operation: 'EntityQueries_CreateOrUpdate' Request Model: 'parameters[5].schema' Response Model: 'responses[200].schema'
> 
> This API has not been changed in the past year. This is an API used by customers and we haven't recieved such errros beforehand, and we cannot change it much.

Please add suppressions for errors on resource types/APIs that are already published in stable API versions (e.g. the above errors you listed, plus a couple AnonymousTypes). See https://aka.ms/azsdk/pr-suppressions for help with suppressions.

Required linter checks that fail on resource types/APIs being added in the current PR need to be fixed. I didn't see any of those, but you should verify.

Note: the BodyTopLevelProperties check is flagging false positives on "value" and "nextLink" properties of collection GET bodies, so those can be ignored. Once suppressions have been added for all the valid errors, reviewers can bypass the false positives.

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2024-04-01-preview/AlertRules.json
Comment (Rotem2dvir) [inline specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2024-04-01-preview/AlertRules.json:784]: These properties are variable and frequently change, with each request potentially having different values for each entity. They were created to accommodate the need for a dynamically changing dictionary, and the data derived from it is utilized across various scenarios.

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2024-04-01-preview/AlertRules.json
Comment (Rotem2dvir) [inline specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2024-04-01-preview/AlertRules.json:1795]: These properties are variable and frequently change, with each request potentially having different values for each entity. They were created to accommodate the need for a dynamically changing dictionary, and the data derived from it is utilized across various scenarios.

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2024-04-01-preview/Recommendations.json
Comment (Rotem2dvir) [inline specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2024-04-01-preview/Recommendations.json:1]: These properties are variable and frequently change, with each request potentially having different values for each entity. They were created to accommodate the need for a dynamically changing dictionary, and the data derived from it is utilized across various scenarios. In office hours , we were not offered different alternative to this use case and we agreed its a valid use case for this name and convention

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2024-04-01-preview/AlertRules.json
Comment (Rotem2dvir) [inline specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2024-04-01-preview/AlertRules.json:1]: These properties are variable and frequently change, with each request potentially having different values for each entity. They were created to accommodate the need for a dynamically changing dictionary, and the data derived from it is utilized across various scenarios. In office hours , we were not offered different alternative to this use case and we agreed its a valid use case for this name and convention

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2024-04-01-preview/TriggeredAnalyticsRuleRuns.json
Comment (Rotem2dvir) [inline specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2024-04-01-preview/TriggeredAnalyticsRuleRuns.json:1]: These properties are variable and frequently change, with each request potentially having different values for each entity. They were created to accommodate the need for a dynamically changing dictionary, and the data derived from it is utilized across various scenarios. In office hours , we were not offered different alternative to this use case and we agreed its a valid use case for this name and convention

**Comment by ramoka178** [comment]
Please fix the lintdiff errors https://github.com/Azure/azure-rest-api-specs/pull/28437/checks?check_run_id=29288998865
There seems to be syntax errors.

**Comment by ramoka178** [comment]
> Please fix the lintdiff errors https://github.com/Azure/azure-rest-api-specs/pull/28437/checks?check_run_id=29288998865 There seems to be syntax errors.

https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/ci-fix.md#swagger-lintdiff-and-swagger-lintrpaas
https://github.com/Azure/azure-openapi-validator/blob/main/CONTRIBUTING.md#how-to-locally-reproduce-a-lintdiff-failure-occurring-on-a-pr

Can help

**Comment by TimLovellSmith** [comment]
All the suppressions look justified to me mostly based on breaking change reasoning.

**Review by viveksheelsingh — APPROVED** [review]

**Pull request merged**
